compiler/rustc_expand/src/expand.rs

```diff
@@ -2179,6 +2179,25 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
         self.visit_node(node)
     }
 
+    fn flat_map_where_predicate(
+        &mut self,
+        mut pred: ast::WherePredicate,
+    ) -> SmallVec<[ast::WherePredicate; 1]> {
+        if self.cx.ecfg.features.cfg_attribute_in_where() {
```

Comment:
The correct way to implement this is to introduce a new `AstFragment`/`Annotatable` for where predicates, and follow what e.g. `flat_map_variant` is doing in general.

Reply:
Also, features should not change the compiler behavior, only report stability error if they are not enabled.

Attributes on where clauses should be feature gated at parsing level, see `psess.gated_spans.gate(...)` for examples.

Reply:
I'd also generally suggest to land the refactorings (`WherePredicateKind`, etc) first, and only then the feature itself (attributes on where predicates, including `cfg`).

Reply:
Thank you for your nice suggestions! I will split this commit into two and land another PR to do that refactoring first.

Reply:
I've opened #132894 for that refactoring.